File: frontend-ci/src/services/histoire.service.ts
/**
 * Service pour les histoires
 * Correspond à gestion/services/histoire_service.py
 */

import apiClient from './api';
import { Histoire, SaveHistoireDTO, UpdateHistoireDTO, CreateResponse } from '../types';

const ENDPOINTS = {
  BASE: '/histoires/',
  DETAIL: (id: number) => `/histoires/${id}/`,
};

export const histoireService = {
  /**
   * Récupérer toutes les histoires (triées par ordre d'affichage)
   */
  async getAll(): Promise<Histoire[]> {
    const response = await apiClient.get<Histoire[]>(ENDPOINTS.BASE);
    return response.data;
  },

  /**
   * Récupérer une histoire par son ID
   */
  async getById(id: number): Promise<Histoire> {
    const response = await apiClient.get<Histoire>(ENDPOINTS.DETAIL(id));
    return response.data;
  },

  /**
   * Créer une nouvelle histoire
   */
  async create(data: SaveHistoireDTO): Promise<CreateResponse> {
    const response = await apiClient.post<CreateResponse>(ENDPOINTS.BASE, data);
    return response.data;
  },

  /**
   * Mettre à jour une histoire
   */
  async update(id: number, data: UpdateHistoireDTO): Promise<Histoire> {
    const response = await apiClient.put<Histoire>(ENDPOINTS.DETAIL(id), data);
    return response.data;
  },

  /**
   * Supprimer une histoire
   */
  async delete(id: number): Promise<void> {
    await apiClient.delete(ENDPOINTS.DETAIL(id));
  },
};

export default histoireService;

File: backend-ci/gestion/services/histoire_service.py
from django.db import transaction
from django.shortcuts import get_object_or_404
from gestion.models.histoire import Histoire


class HistoireService:

    @staticmethod
    @transaction.atomic
    def create(data):
        """Créer une nouvelle histoire"""
        histoire = Histoire.objects.create(
            titre=data['titre'],
            contenu=data['contenu'],
            audio_url=data.get('audio_url', ''),
            image_url=data.get('image_url', ''),
            ordre_affichage=data.get('ordre_affichage', 0)
        )
        return histoire

    @staticmethod
    def get_all():
        """Récupérer toutes les histoires (triées par ordre d'affichage)"""
        return Histoire.objects.all().order_by('ordre_affichage')

    @staticmethod
    def get_by_id(histoire_id):
        """Récupérer une histoire par son ID"""
        return get_object_or_404(Histoire, id=histoire_id)

    @staticmethod
    def update(histoire, data):
        """Mettre à jour une histoire"""
        for key, value in data.items():
            setattr(histoire, key, value)
        histoire.save()
        return histoire

    @staticmethod
    def delete(histoire):
        """Supprimer une histoire"""
        histoire.delete()
